src/commandSocket.cpp

```diff
@@ -447,25 +588,108 @@ void CommandSocket::sendGCodePrefix(std::string prefix)
 #ifdef ARCUS
 std::shared_ptr<cura::proto::Layer> CommandSocket::Private::getLayerById(int id)
 {
-    id += current_layer_offset;
+    id += sliced_layers.current_layer_offset;
 
-    auto itr = sliced_layers.find(id);
+    auto itr = sliced_layers.slice_data.find(id);
 
     std::shared_ptr<cura::proto::Layer> layer;
-    if (itr != sliced_layers.end())
+    if (itr != sliced_layers.slice_data.end())
     {
         layer = itr->second;
     }
     else
     {
         layer = std::make_shared<cura::proto::Layer>();
         layer->set_id(id);
-        current_layer_count++;
-        sliced_layers[id] = layer;
+        sliced_layers.current_layer_count++;
+        sliced_layers.slice_data[id] = layer;
+    }
+
+    return layer;
+}
+#endif
+
+#ifdef ARCUS
+std::shared_ptr<cura::proto::LayerOptimized> CommandSocket::Private::getOptimizedLayerById(int id)
+{
+    id += optimized_layers.current_layer_offset;
+
+    auto itr = optimized_layers.slice_data.find(id);
+
+    std::shared_ptr<cura::proto::LayerOptimized> layer;
+    if (itr != optimized_layers.slice_data.end())
+    {
+        layer = itr->second;
+    }
+    else
+    {
+        layer = std::make_shared<cura::proto::LayerOptimized>();
+        layer->set_id(id);
+        optimized_layers.current_layer_count++;
+        optimized_layers.slice_data[id] = layer;
     }
 
     return layer;
 }
 #endif
 
+#ifdef ARCUS
+void CommandSocket::PathCompiler::flushPathSegments()
+{
+    if( points.size() > 0 && CommandSocket::isInstantiated() )
+    {
+        std::shared_ptr<cura::proto::LayerOptimized> proto_layer = _cs_private_data.getOptimizedLayerById(_layer_nr);
+
+        cura::proto::PathSegment* p = proto_layer->add_path_segment();
+        std::string line_type_data;
+        line_type_data.append(reinterpret_cast<const char*>( line_types.data()), line_types.size()*sizeof(PrintFeatureType) );
+        p->set_line_type(line_type_data);
+        std::string polydata;
+        polydata.append(reinterpret_cast<const char*>( points.data()), points.size() * sizeof(Point));
+        p->set_points(polydata);
+        std::string line_width_data;
+        line_width_data.append(reinterpret_cast<const char*>( line_widths.data()), line_widths.size()*sizeof(int) );
+        p->set_line_width(line_width_data);
+    }
+    points.clear();
+    line_widths.clear();
+    line_types.clear();
+}
+
+//TODO: Reason about these functions when points is empty or the polygon entered is empty
+void CommandSocket::PathCompiler::sendLine(PrintFeatureType print_feature_type, Point from, Point to, int width)
+{
+    handleInitialPoint( from );
```

Comment:
I think it's better to expose this. What I suggest is to rewrite quite a bit of code.

Rather than registering line segments, register moves to a new location, assuming the from-location is the last provided to-location. Besides, gcode also works that way.
Then handle jumps/initial point from outside this class.

So instead of exposing `sendLine(.)`, expose a function `setCurrentLocation(.)` (which is similar to `PathCompiler::handleInitialPoint(.)`) and a function `sendMoveTo(.)`.

Then the functions `GCodePlanner::sendPolygon` (and the one in `MergeInfillLines`) can simply not use the `from` parameter.

The only places where `setCurrentLocation` should be called are:
- at the very start of the gcode it 'jumps' to the first location
- when switching extruder it jumps a relative distance or to an absolute point, based on some machine settings. For the moment I suggest we neglect this case, because it doesn't really matter that the extruder switch shows up as a travel move.